ArduCopter/flight_mode.cpp

```diff
@@ -26,6 +26,11 @@ bool Copter::set_mode(control_mode_t mode, mode_reason_t reason)
         return true;
     }
 
+    // check if an RTL needs to be converted to an AUTO
+    if (mode == RTL && g2.rtl_type == RTLType_MissionDoLandStart) {
+        mode = check_do_land_start();
+    }
+    
```

Comment:
I think this should be before the previous if. We don't want to run auto_init if we are already in auto.
Also I think you added the MODE_REASON_DO_LAND_START to be used here but it is missing (it's not used anywhere).